Remove spacing at start of ImGui uilist and right-align context text.

CLIDragon opened the pull request:
#### Summary
Interface "Improve ImGui uilist spacing and fix text alignment"

#### Purpose of change

Partially fixes #76373

#### Describe the solution

1. Rewrite the separator function to consider its starting point as the same starting point as the text.
2. Hide the unused invisible (well, 4px) columns from the menu. 
3. ~Set `CellPadding.y` to 3, and draw text in columns higher than they would otherwise be.~ Moved to #76420.
4. Right-align the text on the right side of the uilist. Mostly this means removing the excess padding - it was pretty much right-aligned anyway.

~This currently includes #76356, as it was difficult to test windows properly without it. I will remove it before this is ready to merge.~ Rebased and removed.

#### Describe alternatives you've considered

The additional padding could be reduced down to 1px to save screen space.

#### Testing

Unfortunately using the mouse to select filtered text doesn't work. I'm unsure if this is a regression, and if it is, what is causing it.

#### Additional context

There are a few more changes that I would like to make, though they are out of scope for this PR.
1. Increase the border thickness to 2 px. This is the same as the old border size. However, this is nontrivial as ImGui does not natively support borders with a thickness greater than 1 (https://github.com/ocornut/imgui/issues/3138).
2. Fix the issue where the highlighted region moves with the mouse keys out of sync with the selected text.
5. ~Right-align the third column. This is a regression, and one I am working on. It's complicated by the fact that ImGui has no native support for right-aligned text (https://github.com/ocornut/imgui/issues/7805 / https://github.com/ocornut/imgui/issues/7024)~. Done.
6. Properly fix the miscalculated text width instead of just papering over it in `SeparatorEx`.
7. Remove the close button from the window title (and potentiallly replace the entire title with a `TextSeparator`, similar to the old window title).

CLIDragon committed:
Fix blue border, indent in the next section of the text

CLIDragon committed:
Fix length of separator

Brambor commented:
Why not name the PR "Align UI elements and remove inconsistent spacing" ?

CLIDragon commented:
> Why not name the PR "Align UI elements and remove inconsistent spacing" ?

Because I'm not happy with that being the description of what this PR does.

db48x commented on src/cata_imgui.cpp:
```diff
@@ -264,6 +264,8 @@ cataimgui::client::client( const SDL_Renderer_Ptr &sdl_renderer, const SDL_Windo
     // Setup Dear ImGui style
     ImGui::StyleColorsDark();
+    ImGuiStyle &style = ImGui::GetStyle();
+    ImGui::PushStyleVar( ImGuiStyleVar_CellPadding, ImVec2( style.CellPadding.x, 3 ) );
```
Why are you increasing the padding?

db48x commented on src/ui.cpp:
```diff
@@ -132,12 +146,17 @@ void uilist_impl::draw_controls()
                         ImGui::SameLine( 0, 0 );
+                        auto cursorscreenpos = ImGui::GetCursorScreenPos();
+                        ImGui::SetCursorScreenPos( ImVec2( cursorscreenpos.x, cursorscreenpos.y - 2 ) );
+
```
This doesn’t make any sense at all.

CLIDragon replied:
This is an error; it should be 1px.

db48x reviewed (changes requested):
Let’s leave out the mouse and titlebar changes for now, and just concentrate on the other issues.

CLIDragon commented:
The point of both of the above changes is to visually centre the text in the highlighted selection in the ImGui menus. Currently the text appears to be aligned at the bottom of the highlight because in 16px Terminus there is 2 extra pixels of whitespace above the tallest character. Adding at least 1px of padding and then drawing the text 1px higher will visually centre the text.

db48x commented:
Centering the text by changing the cursor position should not be necessary. We would have to do that everywhere we draw text! Even if there is a problem, thatSurely that’s not a good solution.

db48x commented:
Yea, I’ve been experimenting. We can set up the `ImFontConfig` before we load the font, which lets us set a `GlyphOffset`. This changes how the glyphs are positioned in the atlas, and means that we can tweak the glyph positions without munging any padding every time we draw text. For example, here I pushed it a ridiculous amount:

![Screenshot from 2024-09-13 20-24-14](https://github.com/user-attachments/assets/f1127880-fee4-4af7-af23-d3473cc8e77e)

Here’s the font in the Style Editor:

![Screenshot from 2024-09-13 20-27-07](https://github.com/user-attachments/assets/5b23963a-e12f-4986-ad6e-b08067cacefb)

The code, in `cataimgui::client::load_fonts`:

```c++
        ImFontConfig config = ImFontConfig();
        config.GlyphOffset = { 0.0f, -10.0f };
        io.FontDefault = io.Fonts->AddFontFromFileTTF( existing_typeface.c_str(), fontheight, &config,
                         io.Fonts->GetGlyphRangesDefault() );
```

I don’t mind if you include a tweak like this in your patch, but I also note that we have the option of moving to the Freetype render. Freetype can do hinting, which means that small text like this doesn’t end up smeared across so many pixel boundaries. It’ll look better all around once we put in the effort. Plus the game will be smaller by virtue of including only one font rasterizer :)

See also #66749, which updates the game's Freetype renderer to support subpixel anti–aliasing. We ought to revive it while we’re at it.

CLIDragon committed:
Astyle.

CLIDragon commented:
I definitely agree that patch redrawing text is not a good solution. I wasn't even aware that it's possible to change these settings globally. Given the potential scope of this as rewriting the font rendering stack, I have removed the changes, and moved discussion to a new issue (#76420) lest this PR get caught by scope creep.

CLIDragon committed:
Right-align text in the third column of ImGui uilist menus.

CLIDragon commented:
There we go. For future reference, if you want to rebase and your commits diverge from your upstream branch, you should `git branch --unset-upstream <branch_name>` so that git doesn't decide to include all the commits into the branch.

CLIDragon committed:
Fix clang-tidy warnings.

db48x committed:
just turn off NavInputs in uilist windows entirely

Flattening the nav tree solves the blue border issue, but causes other
problems when the user uses the keyboard to wrap the selection around
the top or bottom of the menu.

db48x commented:
I think it looks really good, but I found a way to improve the behavior a bit, so I sent you a pull request of my own.

CLIDragon committed:
Merge pull request #1 from db48x/rebalance-ui

just turn off NavInputs in uilist windows entirely
